### DISCUSSION TITLE
Control - setElement

### DISCUSSION THREAD

**Discussion by tommy1x** [openlayers/openlayers#15330] [Ideas]
Would it be possible to alter Control so it can be created without an element, then set using a method like setElement, same as Overlay/Map, etc.? I'm aware it has setTarget, but it still needs an element to be created as far as I can tell.

This would be helpful as I'm using Svelte, where HTML elements aren't created when the code runs initially. With this change, I could construct a Control before a div is created and then attach once it is:

```
const control = new Control()

const map = getContext("map")
map.addControl(control)

$effect(() => control.setElement(document.getElementById("some-id")))
```

**Comment by ahocevar** [comment]
You can create a `Control` without an element, but to be able to set the element later you have to subclass it:
```js
class MyControl extends Control {
  setElement(element) {
    this.element = element;
  }
}
```
Also make sure you don't call `map.addControl(myContro)` before you have set the `element`.

That said, I don't think you want to use controls for whatever you're trying to achieve. It feels like all you want to do is overlay some markup on the map and let the user interact with it. That doesn't require an OpenLayers `Control`.

**Reply by tommy1x** [reply]
Thank you, very helpful! Is there a reason, though, that Control needs an element to be added but Overlay doesn't?
I was thinking a layer switcher similar to the one on Google Maps, but I guess it may as well be normal markup if I'm not using the Control methods?

**Reply by ahocevar** [reply]
Sorry, I cannot answer this question. However, I guess that's because of the way OpenLayers uses the `Control` base class for its included subclass controls. My advice: Don't use `ol/Control` to create your own controls. Instead, use `setTarget()` on the controls you rely on to move them to a container outside the map, so you can work with them the same way as with other UI elements you overlay on the map.
